pkg/operator/controllers/pullsecret/pullsecret_controller.go

```diff
@@ -75,13 +90,40 @@ func (r *PullSecretReconciler) Reconcile(request ctrl.Request) (ctrl.Result, err
 			ps.Data = map[string][]byte{}
 		}
 
-		// validate
-		if !json.Valid(ps.Data[v1.DockerConfigJsonKey]) {
+		// parse keys and validate JSON
+		parsedKeys, err := r.unmarshalSecretData(ps)
+		foundKey := false
+		failed := false
+		if err != nil {
 			r.log.Info("pull secret is not valid json - recreating")
 			delete(ps.Data, v1.DockerConfigJsonKey)
+			failed = true
+		} else {
+			foundKey = r.checkRHRegistryKeys(parsedKeys)
+		}
+
+		keyCondition := r.keyCondition(failed, foundKey)
+
+		err = controllers.SetCondition(ctx, r.arocli, keyCondition, operator.RoleMaster)
+		if err != nil {
+			return err
+		}
+
+		if cluster.Spec.Features.ManageSamplesOperator {
+			// if foundKey enable samples operator, else disable
+			updated, err := r.switchSamples(ctx, foundKey)
+			if err != nil {
+				return err
+			}
+
+			samplesCondition := r.samplesCondition(updated, foundKey)
+			err = controllers.SetCondition(ctx, r.arocli, samplesCondition, operator.RoleMaster)
+			if err != nil {
+				return err
+			}
```

Comment:
I think all this logic is at the level of complexity where it needs to split. I somehow think we can do a similar boilerplate where functions would be self-contained. This way it will be easier to test. And ran in the sequence they will make sure the order is right. In example, Samples code needs a valid secret, so it makes sense if we fix it before. 
Same with the condition. If redhat secret is present, we set conditions and we can act on it in the next step. 

This might be easier to read and update later. WDYT? 
```
	fm := []func(context.Context){
			r.updateGlobalPullSecret,    // fix & update pull secret if broken
			r.emitRedHatSecretCondition, // updated RedHat pull secret condition
			r.updateSamplesOperator,     // updates samples operator manage state if needed
			r.emitSamplesOperatorCondition,     // updates samples operator condition in Azure operator
		}

		for _, f := range fm {
			err := f(ctx)
			if err != nil {
				return err
			}
		}

```